clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareMembershipManager.java

```diff
@@ -0,0 +1,1105 @@
+import org.apache.kafka.common.metrics.Metrics;
+import org.apache.kafka.common.protocol.Errors;
+import org.apache.kafka.common.requests.ShareGroupHeartbeatRequest;
+import org.apache.kafka.common.telemetry.internals.ClientTelemetryProvider;
+import org.apache.kafka.common.telemetry.internals.ClientTelemetryReporter;
+import org.apache.kafka.common.utils.LogContext;
+import org.apache.kafka.common.utils.Time;
+
+import org.slf4j.Logger;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+/**
+ * Group manager for a single consumer that has a group id defined in the config
+ * {@link ConsumerConfig#GROUP_ID_CONFIG} and the share group protocol to get automatically
+ * assigned partitions when calling the subscribe API.
+ * <p/>
+ *
+ * While the subscribe API hasn't been called (or if the consumer called unsubscribe), this manager
+ * will only be responsible for keeping the member in the {@link MemberState#UNSUBSCRIBED} state,
+ * without joining the group.
+ * <p/>
+ *
+ * If the consumer subscribe API is called, this manager will use the {@link #groupId()} to join the
+ * share group, and based on the share group protocol heartbeats, will handle the full
+ * lifecycle of the member as it joins the group, reconciles assignments, handles fatal errors,
+ * and leaves the group.
+ * <p/>
+ *
+ * Reconciliation process:<p/>
+ * The member accepts all assignments received from the broker, resolves topic names from
+ * metadata, reconciles the resolved assignments, and keeps the unresolved to be reconciled when
+ * discovered with a metadata update. Reconciliations of resolved assignments are executed
+ * sequentially and acknowledged to the server as they complete. The reconciliation process
+ * involves multiple async operations, so the member will continue to heartbeat while these
+ * operations complete, to make sure that the member stays in the group while reconciling.
+ * <p/>
+ *
+ * Reconciliation steps:
+ * <ol>
+ *     <li>Resolve topic names for all topic IDs received in the target assignment. Topic names
+ *     found in metadata are then ready to be reconciled. Topic IDs not found are kept as
+ *     unresolved, and the member request metadata updates until it resolves them (or the broker
+ *     removes it from the target assignment).</li>
+ *     <li>When the above steps complete, the member acknowledges the reconciled assignment,
+ *     which is the subset of the target that was resolved from metadata and actually reconciled.
+ *     The ack is performed by sending a heartbeat request back to the broker.</li>
+ * </ol>
+ *
+ */
+public class ShareMembershipManager implements RequestManager {
```

Comment:
High level thoughts. I expected very similar components (this and the membershipMgr for consumer groups), but after having a first look at this code, don't we have too much in common from the client state machine POV that we're not reusing? (states, transitions, assignment name resolution logic, reconciliation triggering/flow, stale logic). Particularly painful because I see all the recent/inflight tweaks we are right now working on for the membershipMgr, that I see we would end up having to do here again for sure. 

As I see it, the main conceptual differences between the managers for each group would be:

1. HB response processing, given that each has different RPC/data
2. revoke and assign flows, given that it's way simpler for share groups with no callbacks/commit (lucky you).

Am I missing too much of the differences, high level? If not, an option to consider could be keeping a base manager with the core state machine logic that applies to both (I guess we could have a lot here), and have the specific managers overwriting only what they need different. Thoughts? Kind of half-way between having separate components, and having the same component handling both groups.

Reply:
Yes, I agree. I anticipate doing this kind of refactoring, but I was thinking of doing it in a follow-on PR. If you want me to attempt it now, let me know, but it will increase the friction getting the initial merge of the share group consumer code, which is what I was trying to avoid.

The key piece of code is the `ShareConsumeRequestManager`. That's entirely different to the regular consumer. The heartbeating and membership is relatively similar, with a different RPC and subtly different state machine. With some careful work, it would probably be possible to have quite a lot of the heartbeating and membership code in common.

Reply:
Totally fine in as follow-up, just great that we are aligned on the vision.